src/core/progress/progress_manager.cc

```diff
@@ -62,8 +62,9 @@ void progress_manager::finish_work(work* task, bool successfully) {
   tasks.pop();
 
   std::lock_guard<std::mutex> lock(mutex);
-  if (successfully && tasks.empty()) {
-    pbar->set_status_finished();
+  if (tasks.empty()) {
+    if (successfully) pbar->set_status_finished();
+    else              pbar->set_status_error(false);
```

Comment:
I see you did some changes to the progress bar code. What was the reason and do you have any clue as to how [this](http://mr-0xc1:8080/blue/rest/organizations/jenkins/pipelines/h2o-py-data-table/branches/PR-2390/runs/26/nodes/165/log/?start=0) could happen?

```
[2020-04-01T00:20:33.437Z]             assert 0 <= p.progress <= 1
[2020-04-01T00:20:33.437Z]             assert p.status in ("running", "finished", "cancelled", "error")
[2020-04-01T00:20:33.437Z]             progress_reports.append(p)
[2020-04-01T00:20:33.437Z]     
[2020-04-01T00:20:33.437Z]         with dt.options.progress.context(callback = progress_fn,
[2020-04-01T00:20:33.437Z]                                          enabled=True,
[2020-04-01T00:20:33.437Z]                                          min_duration=0):
[2020-04-01T00:20:33.437Z]             [d_exemplars, d_members] = aggregate(d_in, min_rows=0,
[2020-04-01T00:20:33.437Z]                                                  nd_max_bins=div, seed=1)
[2020-04-01T00:20:33.437Z]             messages = ("", "Preparing", "Aggregating", "Sampling", "Finalizing")
[2020-04-01T00:20:33.437Z]             message_index = 0
[2020-04-01T00:20:33.437Z]             for i, p in enumerate(progress_reports):
[2020-04-01T00:20:33.437Z]                 if i > 0:
[2020-04-01T00:20:33.437Z]                     assert p.progress >= progress_reports[i-1].progress
[2020-04-01T00:20:33.437Z]                 message_index_new = messages.index(p.message)
[2020-04-01T00:20:33.438Z] >               assert message_index <= message_index_new or p.status == "finished"
[2020-04-01T00:20:33.438Z] E               AssertionError: assert (1 <= 0 or 'running' == 'finished'
[2020-04-01T00:20:33.438Z] E                 - running
[2020-04-01T00:20:33.438Z] E                 + finished)
```

It looks like the progress bar was still running, but the message it sent was empty. It may not be a big deal since the the same message is normally sent multiple times.

Reply:
The main reason was to have a  well-defined point when the `progress_bar` object in the progress manager should be destroyed. Now it is always destroyed when the last task is popped from the stack. Before we had the progress bar destroyed when the `tasks` becomes empty, but only if the last task finished successfully. If not, then we kept the progress bar instance was kept, and then cleaned up when the exception was caught on the python boundary.
However, this presents a problem if we catch exceptions earlier: basically we have to add some progress-bar-cleanup logic into every exception handler? This just seems utterly error-prone.